pennylane/beta/qnode.py

```diff
@@ -0,0 +1,540 @@
+        self.tape.trainable_params = qml.math.get_trainable_indices(params)
+
+        if not isinstance(self._qfunc_output, Sequence):
+            measurement_processes = (self._qfunc_output,)
+        else:
+            measurement_processes = self._qfunc_output
+
+        if not all(isinstance(m, qml.measure.MeasurementProcess) for m in measurement_processes):
+            raise qml.QuantumFunctionError(
+                "A quantum function must return either a single measurement, "
+                "or a nonempty sequence of measurements."
+            )
+
+        if not all(ret == m for ret, m in zip(measurement_processes, self.tape.measurements)):
+            raise qml.QuantumFunctionError(
+                "All measurements must be returned in the order they are measured."
+            )
+
+        for obj in self.tape.operations + self.tape.observables:
+
+            if getattr(obj, "num_wires", None) is qml.operation.WiresEnum.AllWires:
+                # check here only if enough wires
+                if len(obj.wires) != self.device.num_wires:
+                    raise qml.QuantumFunctionError(
+                        "Operator {} must act on all wires".format(obj.name)
+                    )
+
+    def __call__(self, *args, **kwargs):
+        override_shots = False
+
+        if not self._qfunc_uses_shots_arg:
+            # If shots specified in call but not in qfunc signature,
+            # interpret it as device shots value for this call.
+            override_shots = kwargs.pop("shots", False)
+
+            if override_shots is not False:
+                # Since shots has changed, we need to update the preferred gradient function.
+                # This is because the gradient function chosen at initialization may
+                # no longer be applicable.
+
+                # store the initialization gradient function
+                original_grad_fn = [self.gradient_fn, self.gradient_kwargs, self.device]
+
+                # update the gradient function
+                set_shots(self._original_device, override_shots)(self._update_gradient_fn)()
+
+        # construct the tape
+        self.construct(args, kwargs)
+
+        res = qml.execute(
+            [self.tape],
+            device=self.device,
+            gradient_fn=self.gradient_fn,
+            interface=self.interface,
+            gradient_kwargs=self.gradient_kwargs,
+            override_shots=override_shots,
+            **self.execute_kwargs,
+        )[0]
+
+        if override_shots is not False:
+            # restore the initialization gradient function
+            self.gradient_fn, self.gradient_kwargs, self.device = original_grad_fn
+
+        self._update_original_device()
+
+        if isinstance(self._qfunc_output, Sequence) or (
+            self.tape.is_sampled and self.device._has_partitioned_shots()
+        ):
+            return res
+
+        return qml.math.squeeze(res)
+

```

Comment:
I noticed that all the `to_tf`, `to_torch` methods, etc. are no longer present. Is that now handled somewhere else, or will users just have to set the correct interface/device from the start?

Reply:
It is now handled elsewhere; the interface is passed to `qml.execute(tapes, device, interface)` 🙂 

I think this is much cleaner.

Also, with the `qml.math` module, I think there is a pathway to having tapes infer their interface and trainable parameters on instantiation.

In fact, we would simply need to move this code to the tape:

```python
params = tape.get_parameters(trainable_only=False)
tape.interface = qml.math._multi_dispatch(params)
tape.trainable_params = qml.math.get_trainable_indices(params)
```

(here, `_multi_dispatch` is a -- currently private -- helper function that iterates through a list of mixed types, and determines which is the preferred interface).